src/org/pentaho/reporting/platform/plugin/staging/AsyncJobFileStagingHandler.java

```diff
@@ -0,0 +1,189 @@
+  public static final String STAGING_DIR_ATTR = "asyncstaging";
+
+  private static final Log logger = LogFactory.getLog( AsyncJobFileStagingHandler.class );
+
+  private TrackingOutputStream fileTrackingStream;
+  // package private for testing purpose
+  File tmpFile;
+
+  private static String stringParentDir = ".";
+
+  public AsyncJobFileStagingHandler( OutputStream outputStream, IPentahoSession userSession, String stringParentDir ) throws IOException {
+    // do not write to output stream passed in constructor,
+    // do not keep link to parent output stream, just forget it.
+    // do not to leak.
+    super( null, userSession );
+    this.stringParentDir = stringParentDir;
+  }
+
+  @Override
+  protected void initialize() throws IOException {
+    logger.trace( "Staging mode set - TEMP_FILE, async report generation" );
+    if ( stringParentDir == null ) {
+      throw new IOException( "can't find system/tmp dir, asycn staging not possible." );
+    }
+
+    // /system/tmp/<STAGING_DIR_ATTR>/<session-id>/tmpFile
+    Path stagingExecutionFolder = getStagingExecutionFolder( userSession );
+    if ( !stagingExecutionFolder.toFile().exists() ) {
+      if ( !stagingExecutionFolder.toFile().mkdirs() ) {
+        throw new IOException( "Unable to create staging async directory" );
+      }
+    }
+    Path tempFilePath = stagingExecutionFolder.resolve( UUIDUtil.getUUIDAsString() + POSTFIX );
+    tmpFile = tempFilePath.toFile();
+
+    final Object fileDeleterObj = userSession.getAttribute( ITempFileDeleter.DELETER_SESSION_VARIABLE );
+    if ( fileDeleterObj instanceof ITempFileDeleter ) {
+      ITempFileDeleter fileDeleter = ITempFileDeleter.class.cast( fileDeleterObj );
+      fileDeleter.trackTempFile( tmpFile );
+    } else {
+      logger.debug( "Not found: " + ITempFileDeleter.class.getCanonicalName() );
+      tmpFile.deleteOnExit();
+    }
+    fileTrackingStream = new TrackingOutputStream( new BufferedOutputStream( new FileOutputStream( tmpFile ) ) );
+  }
+
+  private Path getStagingExecutionFolder( IPentahoSession userSession ) {
+    return getStagingDirPath().resolve( userSession.getId() );
+  }
+
+  public static Path getStagingDirPath() {
+    return Paths.get( stringParentDir ).resolve( STAGING_DIR_ATTR );
+  }
+
+  @Override
+  public StagingMode getStagingMode() {
+    return StagingMode.TMPFILE;
+  }
+
+  @Override
+  public boolean isFullyBuffered() {
+    return true;
+  }
+
+  @Override
+  public boolean canSendHeaders() {
+    return true;
+  }
+
+  @Override
+  public OutputStream getStagingOutputStream() {
+    return fileTrackingStream;
+  }
+
+  @Override
+  public void complete() throws IOException {
+    logger.debug(  "Call NO-OP complete: " + this.getClass().getName() );
+  }
+
+  public InputStream getStagingContent() throws FileNotFoundException {
+    return new StagingInputStream( new FileInputStream( tmpFile ) );
+  }
+
+  @Override
+  public void close() {
+    this.closeQuietly();
+  }
+
+  @Override
```

Comment:
Actually, you dont need the TrackingStream at all. Just ask your tmp-file for its file length. This way you also wont hold on to the streams for longer than necessary.